src/goto-instrument/code_contracts.h

```diff
@@ -170,4 +185,175 @@ class code_contractst
 #define HELP_ENFORCE_ALL_CONTRACTS                                             \
   " --enforce-all-contracts      as above for all functions with a contract\n"
 
+/// \brief A base class for assigns clause targets
+class assigns_clause_targett
+{
+public:
+  enum TargetType
+  {
+    Scalar,
+    Struct,
+    Array
+  };
+
+  assigns_clause_targett(
+    TargetType type,
+    const exprt object_ptr,
+    const code_contractst &contract,
+    messaget &log_parameter)
+    : target_type(type),
+      pointer_object(object_ptr),
+      contract(contract),
+      init_block(),
+      log(log_parameter)
+  {
+    INVARIANT(
+      pointer_object.type().id() == ID_pointer,
+      "Assigns clause targets should be stored as pointer expressions.");
+  }
+
+  virtual ~assigns_clause_targett()
+  {
+  }
+
+  virtual std::vector<symbol_exprt> temporary_declarations() const = 0;
+  virtual exprt alias_expression(const exprt &lhs) = 0;
+  virtual exprt
+  compatible_expression(const assigns_clause_targett &called_target) = 0;
+  virtual goto_programt havoc_code(source_locationt location) const = 0;
+
+  const exprt &get_direct_pointer() const
+  {
+    return pointer_object;
+  }
+
+  goto_programt &get_init_block()
+  {
+    return init_block;
+  }
+
+  static const exprt pointer_for(const exprt &exp)
+  {
+    const exprt &left_ptr =
+      exprt(ID_address_of, pointer_type(exp.type()), {exp});
```

Comment:
Use `address_of_exprt` and move this code below the `if` to avoid unnecessary object construction.